apps/fhir/server/authentication.py

```diff
@@ -1,38 +1,125 @@
+        Search the backend FHIR server's patient resource
+        using the mbi_hash identifier.
+    """
+    search_identifier = settings.FHIR_SEARCH_PARAM_IDENTIFIER_MBI_HASH \
+        + "%7C" + mbi_hash
 
-def match_hicn_hash(hicn_hash):
-    auth_state = FhirServerAuth(None)
-    certs = (auth_state['cert_file'], auth_state['key_file'])
-    # URL for patient ID.
-    url = get_resourcerouter().fhir_url + \
-        "Patient/?identifier=http%3A%2F%2Fbluebutton.cms.hhs.gov%2Fidentifier%23hicnHash%7C" + \
-        hicn_hash + \
-        "&_format=json"
+    return search_fhir_id_by_identifier(search_identifier)
+
+
+def search_fhir_id_by_identifier_hicn_hash(hicn_hash):
+    """
+        Search the backend FHIR server's patient resource
+        using the hicn_hash identifier.
+    """
+    search_identifier = settings.FHIR_SEARCH_PARAM_IDENTIFIER_HICN_HASH \
+        + "%7C" + hicn_hash
+
+    return search_fhir_id_by_identifier(search_identifier)
+
+
+def search_fhir_id_by_identifier(search_identifier):
+    """
+        Search the backend FHIR server's patient resource
+        using the specified identifier.
+
+        Return:  fhir_id = matched ID (or None).
+                 err_mesg = None or related error message.
+    """
+    # Get certs from FHIR server settings
+    auth_settings = FhirServerAuth(None)
+    certs = (auth_settings['cert_file'], auth_settings['key_file'])
+
+    # Build URL with patient ID search by identifier.
+    url = get_resourcerouter().fhir_url \
+        + "Patient/?identifier=" + search_identifier \
+        + "&_format=" + settings.FHIR_PARAM_FORMAT
+
+    # Get FHIR service backend response.
+    #   TODO: Should work with verify=True
     response = requests.get(url, cert=certs, verify=False)
     response.raise_for_status()
     backend_data = response.json()
 
-    if backend_data.get('total', 0) > 1:
-        # Don't return a 404 because retrying later will not fix this.
-        raise UpstreamServerException("Duplicate beneficiaries found")
+    # Parse and validate backend_data response.
+    if (
+        'entry' in backend_data
+            and backend_data.get('total', 0) == 1
+            and len(backend_data.get('entry', '')) == 1
+    ):
+        # Found a single matching ID.
+        fhir_id = backend_data['entry'][0]['resource']['id']
+        return fhir_id, None
+    elif (
+        'entry' in backend_data
+            and (backend_data.get('total', 0) > 1 or len(backend_data.get('entry', '')) > 1)
+    ):
+        # Has duplicate beneficiary IDs.
+        return None, "Duplicate beneficiaries found in Patient resource bundle"
+    elif (
+        'entry' in backend_data
+            and backend_data.get('total', -1) == 0
+            and len(backend_data.get('entry', '')) == 0
+    ):
+        # Not found.
+        return None, None
+    else:
+        # Unexpected result! Something weird is happening?
+        return None, "Unexpected result found in the Patient resource bundle"
```

Comment:
> Just curious: why a normal return here instead of raising an exception? (Either one is probably fine, I'm just curious about your philosophy here.)

@karlmdavis You are right!  This did kind of work out in a non standard way and makes the code harder to read!   It is passing the error condition as a message string back to the caller, so that it can be separately included in the audit type log as well as throw the `UpstreamServerException(err_mesg)` up a level.

However, it is more standard to just bubble up the exception and have a handler in the caller `match_fhir_id()` function instead.  I'll refactor the code to do this instead!

Reply:
@karlmdavis I refactored the code per above. I like this better :-)   TY!